lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -0,0 +1,470 @@
+//===- SPIRVToOCL.cpp - Transform SPIR-V builtins to OCL builtins------===//
+//
+//                     The LLVM/SPIRV Translator
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+// Copyright (c) 2014 Advanced Micro Devices, Inc. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal with the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimers.
+// Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimers in the documentation
+// and/or other materials provided with the distribution.
+// Neither the names of Advanced Micro Devices, Inc., nor the names of its
+// contributors may be used to endorse or promote products derived from this
+// Software without specific prior written permission.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
+// THE SOFTWARE.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements common transform of SPIR-V builtins to OCL builtins.
+//
+// Some of the visit functions are translations to OCL2.0 builtins, but they
+// are currently used also for OCL1.2, so theirs implementations are placed
+// in this pass as a common functionality for both versions.
+//
+//===----------------------------------------------------------------------===//
+#define DEBUG_TYPE "spvtocl"
+
+#include "SPIRVToOCL.h"
+
+namespace SPIRV{
+
+static cl::opt<std::string>
+MangledAtomicTypeNamePrefix("spirv-atomic-prefix",
+  cl::desc("Mangled atomic type name prefix"),
+  cl::init("U7_Atomic"));
+
+static cl::opt<unsigned int>
+OCLBuiltinsVersion("spirv-ocl-builtins-version",
+  cl::desc("Specify version of OCL builtins to translate to (ex. 12, 20, 21)"));
```

Comment:
Currently we depend on ocl version specified in llvm metadata opencl.ocl.version. If it's not specified, then getOCLVersion(&M) returns 0 and this leads to applying OCL1.2 builtins translation, because there is the following logic in OCL20to12.cpp:
`if (getOCLVersion(M) >= kOCLVer::CL20)
    return false;`

So, should we set the default value for this option to CL1.2?

Reply:
Okay, then CL1.2 as a default makes sense.